php/libraries/SinglePointLogin.class.inc

```diff
@@ -401,6 +440,84 @@ class SinglePointLogin
         return false;
     }
 
+    /**
+     * Determines whether a user's account is locked due to too many failed
+     * attempts.
+     *
+     * @param string $username The user account to check.
+     *
+     * @return bool
+     */
+    function accountLocked(string $username): bool
+    {
+        /* A user is locked out if they have made X attempts in the last Y
+         * minutes or X' attempts in the last Y' minutes.
+         * The reason for using two time windows is to provide some flexibility:
+         * legitimate locked-out users to try again in a short time while also
+         * preventing attackers from simply trying another X login attempts
+         * after Y minutes.
+         */
+
+        // These represent X and X' above, but with more helpful names.
+        $shortWindowInMinutes = 15;
+        $longWindowInMinutes  = 60;
+        /* These represent Y and Y' above, but with more helpful names.
+         * These values represent the maximum umber of attempts a user may make
+         * to login before being locked out.
+         */
+        $shortLockoutThreshold = 10;
+        $longLockoutThreshold  = 15;
+
+        // SQL code to run.
+        $query  = "SELECT COUNT(loginhistoryID) " .
+            "FROM user_login_history " .
+            "WHERE userID = :username " .
+            "AND Success = 'N' " .
+            "AND IP_address = :ip " .
+            "AND Login_timestamp > ( " .
+            "SELECT IF( MAX(Login_timestamp) > DATE_SUB(NOW(), ".
+            "INTERVAL :timeWindowInMinutes MINUTE), " .
+                "MAX(Login_timestamp), DATE_SUB(NOW(), " .
+                "INTERVAL :timeWindowInMinutes MINUTE) " .
+            ") " .
+            "FROM user_login_history " .
+            "WHERE userID = :username " .
+            "AND Success = 'Y'" .
+        ");";
+        $params = array(
+                   'username'            => $username,
+                   'timeWindowInMinutes' => $shortWindowInMinutes,
+                   'ip'                  => $_SERVER['REMOTE_ADDR'],
+                  );
+
+        $shortFailCount = $this->_getFailedLoginCount($query, $params);
```

Comment:
What I meant was more like:
```
        $database = \NDB_Factory::singleton()->database();
        $statement = $database->prepare($query);
        $shortFailCount = intval($database->execute($statmnent, $params)[0]['COUNT(loginhistoryID)'];
        $params['timeWindowInMinutes'] = $longWindowInMinutes;
        $longFailCount =  intval($database->execute($statmnent, $params)[0]['COUNT(loginhistoryID)'];
```
also, using an alias like 'as loginattempts' in the query would make the result key 'COUNT(loginhistoryID)'  easier to see (subjective comment here)